File: Data Structures & Algorithms/valid-sudoku/submission-3.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        HashSet<Character>[] rows = new HashSet[9]; 
        HashSet<Character>[] cols = new HashSet[9]; 
        HashSet<Character>[] boxes = new HashSet[9]; 

        for(int i = 0;i < 9; i++){
            rows[i] = new HashSet<>();
            cols[i] = new HashSet<>();
            boxes[i] = new HashSet<>();
        }


        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                char c = board[i][j];
                if (c == '.') {
                    continue;
                }
                int boxIndex = (i / 3) * 3 + (j / 3);
                if (rows[i].contains(c) || cols[j].contains(c) || boxes[boxIndex].contains(c)) { // 重复遇到数字 x
                    return false;
                }
                // 标记行、列、宫包含数字 x
                rows[i].add(c);
                cols[j].add(c);
                boxes[boxIndex].add(c);
            }
        }

        return true;
    }
}

File: Data Structures & Algorithms/search-for-word-ii/submission-3.py
class TrieNode:
    __slots__ = ("children", "word")

    def __init__(self):
        self.children: defaultdict[str, TrieNode] = defaultdict(TrieNode)
        self.word: str | None = None


class Solution:

    @staticmethod
    def _buildTrie(words: list[str]) -> TrieNode:
        root = TrieNode()
        for w in words:
            node = root
            for ch in w:
                node = node.children[ch]
            node.word = w
        return root

    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        root = self._buildTrie(words)
        rows, cols = len(board), len(board[0])
        res: list[str] = []
        DIRS = ((1, 0), (-1, 0), (0, -1), (0, 1))

        def dfs(r: int, c: int, node: TriNode) -> None:
            ch = board[r][c]
            if ch == "#" or ch not in node.children:
                return
            nxt = node.children[ch]

            if nxt.word is not None:
                res.append(nxt.word)
                nxt.word = None

            board[r][c] = "#"
            for dr, dc in DIRS:
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols:
                    dfs(nr, nc, nxt)
            board[r][c] = ch
            if not nxt.children and nxt.word is None:
                del node.children[ch]

        for r in range(rows):
            for c in range(cols):
                dfs(r, c, root)
        return res
